code/modules/research/rnd_network_controller.dm

```diff
@@ -0,0 +1,327 @@
+
+	return ..()
+
+
+/obj/machinery/computer/rnd_network_controller/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/computer/rnd_network_controller/ui_interact(mob/user, datum/tgui/ui = null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "RndNetController", name)
+		ui.open()
+
+
+/obj/machinery/computer/rnd_network_controller/ui_data(mob/user)
+	var/list/data = list()
+
+	data["network_password"] = network_password
+	data["network_name"] = network_name
+
+	var/list/devices = list()
+
+	for(var/server_uid in servers)
+		var/obj/machinery/r_n_d/server/S = locateUID(server_uid)
+		if(!S)
+			servers -= server_uid
+			continue
+		devices += list(list("name" = S.name, "id" = S.UID(), "dclass" = "SRV"))
+
+
+	for(var/console_uid in consoles)
+		var/obj/machinery/computer/rdconsole/RDC = locateUID(console_uid)
+		if(!RDC)
+			consoles -= console_uid
+			continue
+		devices += list(list("name" = RDC.name, "id" = RDC.UID(), "dclass" = "RDC"))
+
+
+	for(var/mechfab_uid in mechfabs)
+		var/obj/machinery/mecha_part_fabricator/MF = locateUID(mechfab_uid)
+		if(!MF)
+			mechfabs -= mechfab_uid
+			continue
+		devices += list(list("name" = MF.name, "id" = MF.UID(), "dclass" = "MFB"))
+
+	for(var/backup_uid in backupconsoles)
+		var/obj/machinery/computer/rnd_backup/RB = locateUID(backup_uid)
+		if(!RB)
+			backupconsoles -= backup_uid
+			continue
+		devices += list(list("name" = RB.name, "id" = RB.UID(), "dclass" = "BCK"))
+
+	data["devices"] = devices
+
+	var/list/designs = list()
+
+	// Handle design blacklisting and unblacklisting
+	for(var/datum/design/D in research_files.possible_designs)
+		if(!(D.id in research_files.known_designs) && !(D.id in research_files.blacklisted_designs))
+			continue
+
+		var/list/design_data = list()
+		design_data["name"] = D.name
+		design_data["id"] = D.id
+		design_data["blacklisted"] = (D.id in research_files.blacklisted_designs)
+		design_data["uid"] = D.UID()
+
+		designs += list(design_data)
+
+	data["designs"] = designs
+
+	return data
+
+
+/obj/machinery/computer/rnd_network_controller/ui_act(action, list/params)
+	if(..())
+		return
+
+	. = TRUE
+
+	switch(action)
+		// Set network name
+		if("network_name")
+			var/new_name = tgui_input_text(usr, "Please enter a new network ID", "Network ID", network_name)
+			if(!Adjacent(usr) || !new_name)
+				return
+
+			new_name = trim(new_name)
+
+			if(length(new_name) == 0)
+				return
+
+			if(name_check(new_name))
+				to_chat(usr, "<span class='warning'>Error, network name <code>[new_name]</code> already in use.</span>")
+				return
+
+			to_chat(usr, "<span class='notice'>Network name changed from <code>[network_name]</code> to <code>[new_name]</code>.</span>")
+			network_name = new_name
+
+		if("blacklist_design")
+			var/design_uid = params["d_uid"]
+			var/datum/design/D = locateUID(design_uid)
+
+			// If null design, ignore
+			if(!istype(D, /datum/design))
```

Comment:
Since you cast the variable above, shouldn't you just be able to do `is(!istype(D))` here?